static/app/components/smartSearchBar/index.tsx

```diff
@@ -914,6 +1076,109 @@ class SmartSearchBar extends React.Component<Props, State> {
     );
   }
 
+  /**
+   * Updates autocomplete dropdown items and autocomplete index state
+   *
+   * @param groups Groups that will be used to populate the autocomplete dropdown
+   */
+  updateAutoCompleteStateMultiHeader = (groups: AutocompleteGroup[]) => {
+    const {hasRecentSearches, maxSearchItems, maxQueryLength} = this.props;
+    const {query} = this.state;
+    const queryCharsLeft =
+      maxQueryLength && query ? maxQueryLength - query.length : undefined;
+
+    const searchGroups = groups
+      .map(({searchItems, recentSearchItems, tagName, type}) =>
+        createSearchGroups(
+          searchItems,
+          hasRecentSearches ? recentSearchItems : undefined,
+          tagName,
+          type,
+          maxSearchItems,
+          queryCharsLeft
+        )
+      )
+      .reduce(
+        (acc, item) => ({
+          searchGroups: [...acc.searchGroups, ...item.searchGroups],
+          flatSearchItems: [...acc.flatSearchItems, ...item.flatSearchItems],
+          activeSearchItem: -1,
+        }),
+        {
+          searchGroups: [] as SearchGroup[],
+          flatSearchItems: [] as SearchItem[],
+          activeSearchItem: -1,
+        }
+      );
+
+    this.setState(searchGroups);
+  };
+
+  onAutoCompleteFromAst = (replaceText: string, item: SearchItem) => {
+    const cursor = this.getCursorPosition();
+    const {parsedQuery, query} = this.state;
+    if (!parsedQuery) {
+      return;
+    }
+    const cursorToken = this.getCursorToken(parsedQuery, cursor);
+    if (!cursorToken) {
```

Comment:
still need to handle autocompleting from a default suggestion where the cursor is not on a token. Will do this in a followup pr since this one is getting quite big

Reply:
👍